Documentation / poetry improvements and fixes

sjoerd-bouma opened the pull request:
I suppose technically this could have been two PRs but no functionality should be affected.
1. Optional dependencies weren't really implemented correctly in pyproject.toml (I accept full responsibility). This made very little difference, but the advantage is they can now be installed also using the conventional pip syntax `pip install nuradiomc[documentation, proposal]` to install the correct optional dependencies. Previously, optional dependencies were implemented as developer dependencies, and would never be installed during `pip` installation.
2. The rest is minor fixes to the documentation: some missing docstrings from the recent refactoring, and an attempt to somewhat tidy up the new 'overview of times' page, though I find the content somewhat unclear still. The table of contents on code documentation pages now also only shows top-level submodule and subpackage names instead of every method of every class inside them, which I think improves clarity,

fschlueter commented on documentation/source/NuRadioReco/pages/times.rst:
```diff
@@ -25,28 +25,32 @@ Trace start times in E-fields
+* `NuRadioReco.modules.io.coreas`: CoREAS reader prepends n samples to the simulated trace. This is done so that the trace does not directly start with the pulse and to have a good frequency resolution.
 
-* efieldToVoltageConverter: the traces are rolled (rotated) according to the time delay due to the geometric separation of the antennas and cable delays.
+* `NuRadioReco.modules.efieldToVoltageConverter`: the traces are rolled (rotated) according to the time delay due to the geometric separation of the antennas and cable delays.
```
This isnt this wrong? This module only changes the trace start time but does no rolling.

sjoerd-bouma commented:
ping @fschlueter - I hope you're happy with the changes. I _think_ I have correctly described the way trigger times work as well in times.rst as well as the `trigger` docstrings. If/once we deprecate the triggerTimeAdjuster (#763) its section can be removed.

fschlueter commented on NuRadioReco/framework/trigger.py:
```diff
@@ -103,7 +103,19 @@ def set_trigger_time(self, time):
-        get the trigger time (absolute time with respect to the beginning of the event)
+        Get the trigger time
+
+        Returns the absolute time of the trigger with respect to the beginning of the event,
+        i.e. the first time in the event where the trigger condition was fulfilled
```
The _absolute time_ of the trigger _with respect to_ the beginning of the event sounds like a contradiction by itself to me. If it relative to e.g., the `station_time` or `event_time` I would specify this.

fschlueter commented:
A general comment/question. You are often always omitting the "." at the end of a sentence. I assume this is on purpose? Is there a PEP rule/suggestion on this? Similarly, you do not start sentences with a capitalised letter. Both of that I would do but I do not care much if we just try to be consistent with that in the future (or decide intentionally to be not consistent with that)?

fschlueter commented on documentation/source/NuRadioReco/pages/times.rst:
```diff
@@ -1,52 +1,66 @@
+which is used by both `electric field <NuRadioReco.framework.electric_field.ElectricField>` and `voltage trace <NuRadioReco.framework.channel.Channel>` classes.
+The start of the trace is also accessible separately as the `trace_start_time <NuRadioReco.framework.base_trace.BaseTrace.get_trace_start_time>`.
+
+Time delays are introduced by several hardware components. These time delays are often corrected for by folding/unfolding the complex transfer function (for an amp e.g. the measurement of the S12 parameter).
```
_corrected for_ in this context sounds a bit confusing?

fschlueter commented on documentation/source/NuRadioReco/pages/times.rst:
```diff
@@ -1,52 +1,66 @@
+As a consequence of typically short trace length (<~1000 samples) and because a Fourier transform implies implicitly a periodic signal, a pulse being at the beginning of the trace can end up being at the end of the trace.
+To avoid this behavior we often use the following procedure, through the `NuRadioReco.modules.channelStopFilter` module:
 
 We smoothly filter the first 5% and last 5% of the trace using a Tukey window function. This is a function that goes smoothly from 0 to 1.
-To avoid rollover of the pulse, we add 128ns of zeros to the beginning and end of the trace. Steps 2) and 3) are performed by the channelStopFilter module
-Both electric fields and channels have a trace_start_time variable. The get_times() function will automatically add the trace_start_time to the time array.
+To avoid rollover of the pulse, we add 128ns of zeros to the beginning and end of the trace.
```
Is that really true?

fschlueter replied:
Yes, but how it is written within the docu suggests that it is the common/default behaviour. Maybe we can add this module to the list of modules which change the timing?

sjoerd-bouma replied:
It is/was already there. As an aside, it doesn't do what it says here (the trace start time is not adjusted), but I will make a separate bugfix for that.

fschlueter commented on documentation/source/NuRadioReco/pages/times.rst:
```diff
@@ -1,52 +1,66 @@
-The trace_start_times are all given relative to the station_time of the station the E-field or channel belongs to. The station_time is stored in an astopy.time.Time object for sub nanosecond precision on absolute times.
-The trace_start_time itself is stored as a float. For simulations, the trace_start_time is relative to the vertex time, i.e., the time of the particle interaction. 
+
+The `trace_start_time <NuRadioReco.framework.base_trace.BaseTrace.get_trace_start_time>` is always relative to the station_time of the station the E-field or channel belongs to. The station_time is stored in an `astropy.time.Time` object for sub nanosecond precision on absolute times.
```
link to `station_time`

sjoerd-bouma commented:
> A general comment/question. You are often always omitting the "." at the end of a sentence. I assume this is on purpose? Is there a PEP rule/suggestion on this? Similarly, you do not start sentences with a capitalised letter. Both of that I would do but I do not care much if we just try to be consistent with that in the future (or decide intentionally to be not consistent with that)?

I agree. Though I would like to point out that `times.rst` was not written by me and I'm just trying to clean it up a bit. I also don't think it's worth getting too upset over typos, missing capitals or full stops in doc strings (and I don't think it's necessary to enforce some kind of standard), though obviously correctly written and formatted text is nicer to read.

fschlueter commented:
> > A general comment/question. You are often always omitting the "." at the end of a sentence. I assume this is on purpose? Is there a PEP rule/suggestion on this? Similarly, you do not start sentences with a capitalised letter. Both of that I would do but I do not care much if we just try to be consistent with that in the future (or decide intentionally to be not consistent with that)?
> 
> I agree. Though I would like to point out that `times.rst` was not written by me and I'm just trying to clean it up a bit. I also don't think it's worth getting too upset over typos, missing capitals or full stops in doc strings (and I don't think it's necessary to enforce some kind of standard), though obviously correctly written and formatted text is nicer to read.

I was not referring to `times.rst' but the doc-strings in function definitions.

fschlueter commented on pyproject.toml:
```diff
@@ -42,25 +42,23 @@ uproot = "*"
 galacticnoise = ['pygdsm']
 ift_reco = ['nifty5', 'pypocketfft']
-muon_flux_calc = ['MCEq', 'crflux']
-RNO_G_DATA = ["mattak", "runtable", "pandas"]
```
Why removing runtable and pandas from the RNO_G_DATA dependencies?

sjoerd-bouma replied:
I've added a separate `rno-g-extras` option now with the runtable, that does not get installed by default. Not sure if there is a better solution; in any case once you try to use the RNO-G reader it suggests installing the run table, I think.

fschlueter commented on documentation/source/NuRadioReco/pages/times.rst:
```diff
@@ -1,31 +1,48 @@
+Times in `Channel <NuRadioReco.framework.channel.Channel>`, `ElectricField <NuRadioReco.framework.electric_field.ElectricField>` and
+`Trigger <NuRadioReco.framework.trigger.Trigger>` objects are all defined relative to the
+`station_time <NuRadioReco.framework.station.Station.get_station_time>` of the Station they are stored in (see the description
+of the :doc:`NuRadio data structure </NuRadioReco/pages/event_structure>`). These times are stored as an array of floats.
```
>  These times are stored as an array of floats.

To what does this refer?

fschlueter commented on documentation/source/NuRadioReco/pages/times.rst:
```diff
@@ -1,31 +1,48 @@
+
+Station time (Event time)
+-------------------------
+The global time at which the event takes place is stored as the `event time <NuRadioReco.framework.event.Event.get_event_time>`
+in the `Event <NuRadioReco.framework.event.Event>` object.
+This time usually corrsponds to the "vertex time" of the first interaction for simulations,
```
For data the event and station time are typically the same, right? The only exception I can think of is that we - as part of an event reconstruction - merge events across stations. Can we add a statement here - just to be very explicit.

fschlueter commented:
@sjoerd-bouma great work, thank you so much! I have a few more minor comments (I think we should make the documentation explicit AF to never touch it again (or at least not for a very long time).

fschlueter commented on NuRadioReco/detector/RNO_G/db_mongo_read.py:
```diff
@@ -720,7 +735,38 @@ def get_channel_signal_chain_measurement(self, station_id=None, channel_id=None,
 
     def get_component_data(self, component_type, component_id, supplementary_info, primary_time, verbose=True, sparameter='S21'):
-        """ returns the current primary measurement of the component, reads in the component collection"""
+        """
+        returns the current primary measurement of the component, reads in the component collection
```
Minor comment, I think with the PR #788 I am adding a wrapper around this function within the detector class which probably should be used by users.

sjoerd-bouma replied:
Then maybe you can update this docstring / the module docstring in #788? Feel free to copy and paste stuff from here if you think it applies.

fschlueter commented on documentation/source/NuRadioReco/pages/times.rst:
```diff
@@ -1,52 +1,95 @@
+  If ``sim to data=True``:
 
-* hardwareResponseIncorporator (sim to data):
     * the channel traces are folded with the amplifier response which also includes some time delay
```
I would add a statement that this time delay is reflected by moving the pulse within the trace and not by adjusting the trace_start_time

fschlueter commented on documentation/source/NuRadioReco/pages/times.rst:
```diff
@@ -1,52 +1,95 @@
+in the `Event <NuRadioReco.framework.event.Event>` object.
+This time usually corrsponds to the "vertex time" of the first interaction for simulations,
+and the time at which the data was recorded in the DAQ for data.
+It is stored as an `astropy.time.Time` object to enable sub-ns precision on the absolute time.
```
Hm that is only true for data, right? In simulations this is typically also "just" a float. Should we change that?

sjoerd-bouma commented:
@fschlueter hopefully I've addressed your remaining comments. I also reformatted a couple of docstrings that were causing warnings during compilation.
